RFC/src/RFC-0171_MessageSerialisation.md

```diff
@@ -0,0 +1,175 @@
+
+This document only covers the first and last steps: _viz_: serialising data from in-memory objects to a format that can
+be transmitted over the wire. The other steps are handled by the Tari communication protocol.
+
+In addition to machine-to-machine communication, we also standardise on human-to-machine communication. Use cases for
+this include
+
+* Handcrafting instructions or transactions. The ideal format here is a very human-readable format.
+* Copying transactions or instructions from cold wallets. The ideal format here is a compact but easy-to-copy format.
+* Peer-to-peer text messaging. This is just a special case of what has already been described, with the message
+  structure containing a unicode `message_text` field.
+
+When sending a message from a human to the network, the following happens:
+
+* The message is deserialised into the native structure.
+* The deserialisation acts as an automatic validation step.
+* Additional validation can be performed.
+* The usual machine-to-machine process is followed as described above.
+
+### Binary serialisation formats
+
+The ideal properties for binary serialisation formats are:
+
+* Widely used across multiple platforms and languages, but with excellent Rust support.
+* Compact binary representation
+* Serialisation "Just Works"(TM) with little or no additional coding overhead.
+
+Several candidates fulfil these properties to some degree.
+
+#### [ASN.1](http://www.itu.int/ITU-T/asn1/index.html)
+
+* Pros:
+  * Very mature (was developed in the 80s)
+  * Large number of implementations
+  * Dovetails into ZMQ nicely
+* Cons:
+  * Limited Rust / Serde support
+  * Requires schema (additional coding overhead if no automated tools for this exist)
+
+
+#### [Message Pack](http://msgpack.org/)
+
+* Pros:
+  * Very compact
+  * Fast
+  * Multiple language support
+  * Good Rust / SerDe support
+  * Dovetails into ZMQ nicely
+* Cons:
+  * No metadata support
+
+#### [Protobuf](https://code.google.com/p/protobuf/)
+
+Similar to [Message Pack](#message-pack), but also requires schema's to be written and compiled. Serialisation performance and size
+is similar to Message Pack. Can work with ZMQ but is better designed to be used with gRPC.
+
+#### [Cap'n Proto](http://kentonv.github.io/capnproto/)
+
+Similar to [Protobuf](#protobuf), but claims to be much faster. Rust is supported.
+
+### Serialisation in Tari
+
+Deciding between these protocols is largely a matter of preference, since there isn't that much to choose between them.
+Given that ZMQ is used in other places in the Tari network; MessagePack looks to be a good fit while offering a compact
+data structure and highly performant de/serialisation. In Rust in particular, there's first-class support for Message
+Pack in serde.
+
+For human-readable formats, it makes little sense to deviate from JSON. For copy-paste semantics, the extra compression
+that Base64 offers over raw hex or Base58 makes it attractive.
```

Comment:
Does base64 really create that much better compression over base58 in that the compromise is worth while? In the JSON format we are going to human readability and base58 is much clearer in that.

Reply:
Base64 is more for encoding into QR codes and copy-paste; so it's not meant to be human-readable, just ASCII-friendly. Base64 lines up on word boundaries better, in addition to squeezing a bit more info into every character. From [Wikipedia](https://en.wikipedia.org/wiki/Base58):

>  In contrast to Base64, the digits of the [Base58] encoding do not line up well with byte boundaries of the original data. For this reason, the method is well-suited to encode large integers, but not designed to encode longer portions of binary data.

The only real drawback to Base64 is that a double-click won't select the whole string if there are `+` or `/` in it